### PR TITLE
Strong type python functions

### PR DESCRIPTION
**Pull request by royico**
### What does this PR do?
Adding Strong type to function parameters
Strong typing means that variables do have a type and that the type matters when performing operations on a variable

### What ticket does this PR close?
https://app.zenhub.com/workspaces/palmtree-5d99d900491c060001c85cba/issues/cyberark/conjur-api-python3/193

#### Change log
- [ ] The CHANGELOG has been updated, or
- [X] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [X] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [X] This PR does not require updating any documentation

### CODE REVIEW

File: conjur/api/api.py [hunk 1]
```diff
@@ -20,13 +20,11 @@
 # pylint: disable=too-many-instance-attributes
 from conjur.resource import Resource
```
Comment (mbenita-Cyberark) [inline conjur/api/api.py:22]: why delete all the new lines? is it done by auto-ident?

File: conjur/api/client.py [hunk 2]
```diff
@@ -49,15 +48,15 @@ class Client():
     # what parameters are allowed
     # pylint: disable=too-many-arguments,too-many-locals,too-many-branches,line-too-long,try-except-raise,too-many-statements
     def __init__(self,
-                 account=None,
+                 account:str=None,
                  api_key=None,
                  ca_bundle=None,
```
Comment (mbenita-Cyberark) [inline conjur/api/client.py:53]: what about api_key and ca_bundle?  they are both strings

File: conjur/api/ssl_client.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 import socket
 from OpenSSL import SSL
 from OpenSSL.crypto import FILETYPE_PEM, dump_certificate
+import inspect
```
Comment (mbenita-Cyberark) [inline conjur/api/ssl_client.py:14]: what is inspect?

File: conjur/logic/variable_logic.py [hunk 4]
```diff
@@ -18,6 +19,7 @@ class VariableLogic:
     This class holds the business logic for executing and manipulating
     returned data
     """
+
     def __init__(self, client):
```
Comment (mbenita-Cyberark) [inline conjur/logic/variable_logic.py:23]: what about strong typing here?

File: conjur/wrapper/http_wrapper.py [hunk 5]
```diff
@@ -32,8 +32,8 @@ class HttpVerb(Enum):
 #pylint: disable=too-many-locals
 # ssl_verify can accept Boolean or String as per requests docs
 # https://requests.readthedocs.io/en/master/api/#main-interface
-def invoke_endpoint(http_verb, endpoint, params, *args, check_errors=True,
-                    ssl_verify=True, auth=None, api_token=None, query=None):
+def invoke_endpoint(http_verb:HttpVerb, endpoint, params:dict, *args, check_errors:bool=True,
```
Comment (mbenita-Cyberark) [inline conjur/wrapper/http_wrapper.py:35]: endpoint should be "ConjurEndpoint" enum

**Review by mbenita-Cyberark — COMMENTED** [review]
Looks great.

Added comments.
please see that constructors are types as well

File: conjur/api/api.py (hunk 1 above)
Comment (royico) [reply conjur/api/api.py:22]: By Mistake, when i removed my checking function
Fixed.

File: conjur/api/client.py (hunk 2 above)
Comment (royico) [reply conjur/api/client.py]: done

File: conjur/api/ssl_client.py (hunk 3 above)
Comment (royico) [reply conjur/api/ssl_client.py]: leftovers , Thanks

File: conjur/api/api.py [hunk 6]
```diff
@@ -79,7 +79,7 @@ def __init__(self,
 
     @property
     # pylint: disable=missing-docstring
-    def api_token(self):
+    def api_token(self) -> requests.Response:
```
Comment (mbenita-Cyberark) [inline conjur/api/api.py:82]: pretty sure it is str. but i could be wrong

Comment (mbenita-Cyberark) [reply conjur/api/api.py:82]: yep it is str

File: conjur/api/api.py [hunk 7]
```diff
@@ -109,12 +109,13 @@ def login(self, login_id=None, password=None):
 
         return self.api_key
 
-    def authenticate(self):
+    def authenticate(self) -> requests.Response:
```
Comment (mbenita-Cyberark) [inline conjur/api/api.py:112]: return str 
look it return invoke_endpoint(HttpVerb.POST, ConjurEndpoint.AUTHENTICATE, params,
                               self.api_key, ssl_verify=self._ssl_verify)**.text**

**Review by mbenita-Cyberark — CHANGES_REQUESTED** [review]
will continue later

File: conjur/api/api.py (hunk 7 above)
Comment (royico) [reply conjur/api/api.py]: Done

File: conjur/api/api.py (hunk 6 above)
Comment (royico) [reply conjur/api/api.py:82]: Done

**Review by mbenita-Cyberark — APPROVED** [review]
LGTM

File: conjur/api/api.py [hunk 8]
```diff
@@ -156,13 +152,14 @@ def resources_list(self, list_constraints=None):
         if list_constraints is not None and 'inspect' not in list_constraints:
             # For each element (resource) in the resources sequence, we extract the resource id
             resource_list = map(lambda resource: resource['id'], resources)
+            # TODO: Understand why list and not dict
```
Comment (sigalsax) [inline conjur/api/api.py:155]: @royico did we understand why this isn't a dict? :)

File: conjur/argument_parser/_host_parser.py [hunk 9]
```diff
@@ -71,7 +71,7 @@ def _host_rotate_api_key_parser(sub_parser):
                                          help='Display help screen and exit')
 
     @staticmethod
-    def _add_host_options(host_subparser):
+    def _add_host_options(host_subparser:ArgparseWrapper):
```
Comment (sigalsax) [inline conjur/argument_parser/_host_parser.py:74]: missing space after `:`

File: conjur/argument_parser/_host_parser.py [hunk 10]
```diff
@@ -47,7 +47,7 @@ def _create_host_parser(self):
         return host_subparser
 
     @staticmethod
-    def _host_rotate_api_key_parser(sub_parser):
+    def _host_rotate_api_key_parser(sub_parser:ArgparseWrapper):
```
Comment (sigalsax) [inline conjur/argument_parser/_host_parser.py:50]: missing space after `:`

File: conjur/cli.py [hunk 11]
```diff
@@ -129,7 +132,7 @@ def handle_init_logic(cls, url=None, account=None, cert=None, force=None, ssl_ve
 
     @classmethod
     # pylint: disable=line-too-long
-    def handle_login_logic(cls, credential_provider, identifier=None, password=None, ssl_verify=True):
+    def handle_login_logic(cls, credential_provider:CredentialsStoreInterface, identifier:str=None, password:str=None, ssl_verify:bool=True):
```
Comment (sigalsax) [inline conjur/cli.py:135]: nit missing space after `:`

File: conjur/controller/init_controller.py [hunk 12]
```diff
@@ -138,14 +143,15 @@ def get_account_info(self, conjurrc_data):
                 # a 401 status code will be returned.
                 # If the endpoint does not exist, the user will be prompted to enter in their account.
                 # pylint: disable=no-member
+                # TODO: If respone not exist in error then we will have a ecxption here
```
Comment (sigalsax) [inline conjur/controller/init_controller.py:146]: why this done?

**Comment by sigalsax** [comment]
a little late to the party but added a few comments :)

**Pull request merged**